### ISSUE TITLE
Dead?

### ISSUE THREAD

**Issue by goodmind** [dot-build/type-language#1]
Are you still working on this?

**Comment by darlanalves** [comment]
Hi @goodmind!

At the moment, no. I started the project as a weekend project to have a decent implementation of it in JS, but got carried away with another priorities at work unfortunately 😞 

There is an initial work done in a [separated branch](https://github.com/dot-build/type-language/tree/v1), though. If you have interest in contribute, please ping back 😄 

Cheers,
Darlan.

**Comment by goodmind** [comment]
Is this implementation compatible with https://github.com/enricostara/telegram-tl-node (or will be?)

**Comment by darlanalves** [comment]
Nope, is actually an effort to complete replace this library. I wrote an app using these libraries and they are completely unreadable, far from a clean code and buggy sometimes. I started another library as a replacement for telegram.link at first, then I rewrote the entire mt-node one to use in a project and this one was the last piece of the puzzle.

Unfortunately I left this project and joined another one in a completely different area. Since then I didn't get a major progress on this one.

A friend o'mine, tho, is currently working on a new version of tl-node as well, but is on a private project 😞

**Comment by goodmind** [comment]
@darlanalves where is rewrote of mt-node? is it private?

**Comment by darlanalves** [comment]
Yep. I essencially rewrote all the libraries to some extent for a private project. While studying the libraries to understand Telegram I created this project in parallel to use as a generic codec in other projects.

**Comment by goodmind** [comment]
@darlanalves can you open source it?

**Comment by darlanalves** [comment]
Unfortunately nope. Is part of a closed source project for a company.
Beside that, my rewrite was still using lots of pieces for encryption that were based on the mt-node library. The code is not that great anyways. A better option would be start an effort to write a decent and well designed library for both MTProto and the TL language.

This project started as a step towards that direction, but is very complex to cover the full spectrum of things to put together for Telegram API, specially the gazillion of crypto-related libraries required to make it work.

If you need a reference code to work upon, I'd recommend this app:
https://github.com/actorapp

Actor was founded by a guy that used to work for Telegram. The mt-node library is a Node.JS port of another library which was written in Java by this guy as well.

Or just give up and run to the hills. This entire thing is insane 🤣

**Comment by goodmind** [comment]
@darlanalves Hey, take a look at https://github.com/zerobias/telegram-mtproto/tree/mtproto2
This is port of webogram's code (lol yes)

**Comment by darlanalves** [comment]
Hi @goodmind !

Nice effort on it :) ES6+ makes it soo much easier to read.

Webogram has a different approach on it and seems a bit more reasonable in terms of code structure. The TL implementation they use is a bit more readable as well.

What made me run away as fast as possible from telegram-tl-node was this:
https://github.com/zerobias/telegram-mtproto/blob/master/lib/tl/builder/constructor-builder.js

The TL language is not so easy to understand and implement. I've made [an initial version of it](https://github.com/dot-build/type-language/tree/v1) to test a concept but never finished it unfortunately. The main problems with the libraries we have so far are huge dependencies on crypto modules and bad design. I want to come back to this library one day and finish it as a general-purpose serialization/deserialization lib for WebSockets. The communication time using binary protocols is amazing, specially on realtime apps like games or IM.

Keep on going in your version! The community needs a neat library for Telegram!
